### PR TITLE
Add aws codebuild source credential resource

### LINKED ISSUE #7025: AWS Codebuild Import Source Credentials
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
This feature will add the ability to connect using a token with Bitbucket and Github so the provisioned projects have the Oauth token available already. It should eliminate the manual step of connecting codebuild project to the respective SCM.

### New or Affected Resource(s)

* aws_codebuild_project
* aws_codebuild_webhook

### References

* AWS API - https://docs.aws.amazon.com/codebuild/latest/APIReference/API_ImportSourceCredentials.html
* AWS API - https://docs.aws.amazon.com/codebuild/latest/APIReference/API_DeleteSourceCredentials.html

### LINKED ISSUE #7435: aws_codebuild_project: support github token
It would be great if aws_codebuild_project could support github `auth` via github token:

screenshot 2019-02-03 at 21 58 53

### PR DESCRIPTION
**Pull request by kterada0509**
Closes #7025
Closes #7435

Changes proposed in this pull request:

* Add `aws_codebuild_source_credential` resource

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSCodeBuildSourceCredential_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSCodeBuildSourceCredential_ -timeout 120m
=== RUN   TestAccAWSCodeBuildSourceCredential_Basic
=== PAUSE TestAccAWSCodeBuildSourceCredential_Basic
=== RUN   TestAccAWSCodeBuildSourceCredential_BasicAuth
=== PAUSE TestAccAWSCodeBuildSourceCredential_BasicAuth
=== CONT  TestAccAWSCodeBuildSourceCredential_Basic
=== CONT  TestAccAWSCodeBuildSourceCredential_BasicAuth
--- PASS: TestAccAWSCodeBuildSourceCredential_Basic (49.43s)
--- PASS: TestAccAWSCodeBuildSourceCredential_BasicAuth (49.55s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	49.623s
```

### CODE REVIEW

**Comment by kterada0509** [comment]
Re-run acctest.

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSCodeBuildSourceCredential_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSCodeBuildSourceCredential_ -timeout 120m
=== RUN   TestAccAWSCodeBuildSourceCredential_Basic
=== PAUSE TestAccAWSCodeBuildSourceCredential_Basic
=== RUN   TestAccAWSCodeBuildSourceCredential_BasicAuth
=== PAUSE TestAccAWSCodeBuildSourceCredential_BasicAuth
=== CONT  TestAccAWSCodeBuildSourceCredential_Basic
=== CONT  TestAccAWSCodeBuildSourceCredential_BasicAuth
--- PASS: TestAccAWSCodeBuildSourceCredential_Basic (53.06s)
--- PASS: TestAccAWSCodeBuildSourceCredential_BasicAuth (53.12s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	53.212s
```

File: aws/resource_aws_codebuild_source_credential.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+func resourceAwsCodeBuildSourceCredential() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsCodeBuildSourceCredentialCreate,
+		Read:   resourceAwsCodeBuildSourceCredentialRead,
+		Delete: resourceAwsCodeBuildSourceCredentialDelete,
+
```
Comment (bflad) [inline aws/resource_aws_codebuild_source_credential.go]: Can you please implement resource import support? More information on the required steps can be found in the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#adding-resource-import-support). This resource testing will likely require `ImportStateVerifyIgnore: []string{"token", "user_name"}` Thanks!

```suggestion
		Importer: &schema.ResourceImporter{
			State: schema.ImportStatePassthrough,
		},

```

File: aws/resource_aws_codebuild_source_credential.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	authType := d.Get("auth_type").(string)
+
+	createOpts := &codebuild.ImportSourceCredentialsInput{
+		AuthType:   aws.String(authType),
+		ServerType: aws.String(d.Get("server_type").(string)),
+		Token:      aws.String(d.Get("token").(string)),
+	}
+
+	if attr, ok := d.GetOk("user_name"); ok && attr.(string) != "" && authType == codebuild.AuthTypeBasicAuth {
```
Comment (bflad) [inline aws/resource_aws_codebuild_source_credential.go]: `d.GetOk()` automatically checks for zero values so the empty string check is extraneous 👍 

 ```suggestion
	if attr, ok := d.GetOk("user_name"); ok && authType == codebuild.AuthTypeBasicAuth {
```

File: aws/resource_aws_codebuild_source_credential.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		createOpts.Username = aws.String(attr.(string))
+	}
+
+	resp, err := conn.ImportSourceCredentials(createOpts)
+	if err != nil {
+		return fmt.Errorf("Error importing source credentials: %s", err)
+	}
+
+	d.SetId(aws.StringValue(resp.Arn))
+	d.Set("arn", resp.Arn)
```
Comment (bflad) [inline aws/resource_aws_codebuild_source_credential.go]: Preferably the `Create` function should return the `Read` function and this `d.Set()` should be handled in there. 👍 

```suggestion
```

File: aws/resource_aws_codebuild_source_credential.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+	return nil
```
Comment (bflad) [inline aws/resource_aws_codebuild_source_credential.go]: Preferably the `Create` function should return the `Read` function.

```suggestion
	return resourceAwsCodeBuildSourceCredentialRead(d, meta)
```

File: aws/resource_aws_codebuild_source_credential.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+		return fmt.Errorf("Error list source credentials: %s", err)
+	}
+
+	if len(resp.SourceCredentialsInfos) == 0 {
+		log.Printf("[WARN] Source Credentials(%s) is already deleted", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	resourceNotFound := true
```
Comment (bflad) [inline aws/resource_aws_codebuild_source_credential.go]: Typically we prefer to use relevant object type as the output variable when searching like this, which simplifies the logic and allows the code to read well on the left side:

```go
var info *codebuild.SourceCredentialInfo

for _, sourceCredentialsInfo := range resp.SourceCredentialsInfos {
	if d.Id() == aws.StringValue(sourceCredentialsInfo.Arn) {
		info = sourceCredentialInfo
		break
	}
}

if info == nil {
	log.Printf("[WARN] CodeBuild Source Credential (%s) not found, removing from state", d.Id())
	d.SetId("")
	return nil
}

d.Set("arn", info.Arn)
d.Set("auth_type", info.AuthType)
d.Set("server_type", info.ServerType)

return nil
```

File: aws/resource_aws_codebuild_source_credential.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+func resourceAwsCodeBuildSourceCredentialDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).codebuildconn
+
+	deleteOpts := &codebuild.DeleteSourceCredentialsInput{
+		Arn: aws.String(d.Id()),
+	}
+
+	if _, err := conn.DeleteSourceCredentials(deleteOpts); err != nil {
+		if !isAWSErr(err, codebuild.ErrCodeResourceNotFoundException, "") {
```
Comment (bflad) [inline aws/resource_aws_codebuild_source_credential.go]: Typo 😎 

```suggestion
		if isAWSErr(err, codebuild.ErrCodeResourceNotFoundException, "") {
```

File: website/docs/r/codebuild_source_credential.html.markdown [hunk 7]
```diff
@@ -0,0 +1,48 @@
+---
+layout: "aws"
+page_title: "AWS: aws_codebuild_source_credential"
+sidebar_current: "docs-aws-resource-codebuild-source-credential"
+description: |-
+  Provides a CodeBuild Source Credential resource.
+---
+
+# aws_codebuild_project
```
Comment (bflad) [inline website/docs/r/codebuild_source_credential.html.markdown]: Copy pasta 🍝 and sitewide documentation update that occurred after this PR was submitted 👍 

```suggestion
# Resource: aws_codebuild_source_credential
```

Comment (kterada0509) [reply website/docs/r/codebuild_source_credential.html.markdown]: fixed in ca65197

File: aws/resource_aws_codebuild_source_credential.go (hunk 1 above)
Comment (kterada0509) [reply aws/resource_aws_codebuild_source_credential.go]: fixed in c13110c

File: aws/resource_aws_codebuild_source_credential.go (hunk 6 above)
Comment (kterada0509) [reply aws/resource_aws_codebuild_source_credential.go]: fixed in 588a1c3

File: aws/resource_aws_codebuild_source_credential.go (hunk 5 above)
Comment (kterada0509) [reply aws/resource_aws_codebuild_source_credential.go]: fixed in 588a1c3

File: aws/resource_aws_codebuild_source_credential.go (hunk 4 above)
Comment (kterada0509) [reply aws/resource_aws_codebuild_source_credential.go]: fixed in 05104d7

File: aws/resource_aws_codebuild_source_credential.go (hunk 3 above)
Comment (kterada0509) [reply aws/resource_aws_codebuild_source_credential.go]: fixed in de6da2b

File: aws/resource_aws_codebuild_source_credential.go (hunk 2 above)
Comment (kterada0509) [reply aws/resource_aws_codebuild_source_credential.go]: fixed in 64959a1

**Comment by kterada0509** [comment]
Re-run acctest.

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSCodeBuildSourceCredential_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSCodeBuildSourceCredential_ -timeout 120m
=== RUN   TestAccAWSCodeBuildSourceCredential_Basic
=== PAUSE TestAccAWSCodeBuildSourceCredential_Basic
=== RUN   TestAccAWSCodeBuildSourceCredential_BasicAuth
=== PAUSE TestAccAWSCodeBuildSourceCredential_BasicAuth
=== CONT  TestAccAWSCodeBuildSourceCredential_Basic
=== CONT  TestAccAWSCodeBuildSourceCredential_BasicAuth
--- PASS: TestAccAWSCodeBuildSourceCredential_Basic (51.97s)
--- PASS: TestAccAWSCodeBuildSourceCredential_BasicAuth (52.19s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	52.268s
```

@bflad ready for re-review.

**Pull request merged**